daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DPathRuntime.scala

```diff
@@ -95,7 +99,7 @@ class CompiledDPath(val ops: RecipeOp*) extends Serializable {
         // in a way that is not consistent with a constant expression. Such as trying to do
         // anything with the infoset other than saving and restoring current position in the infoset.
         // Ditto trying to read a variable.
-        case _: InfosetException | _: VariableException | _: java.lang.IllegalStateException =>
+        case _: InfosetException | _: VariableException | _: NoSuchElementException | _: java.lang.IllegalStateException =>
```

Comment:
What causes this NoSuchElementException? I thought we only want to catch Daffodil specific exceptions here, like the infoset stuff or variable stuff? NoSuchELementException implies to me that we have a bug somewhere and are accessing somethign we shouldn't.

Reply:
dfdlx:repTypeValue() and dfdlx:logicalTypeValue() attempt to read the current rep/logical value from a Maybe variable. At compile time, these are always Nope, and so throw a NoSuchElementException.

There is a good arguement to be made that we shouldn't catch NoSuchElementHere, as this will also catch bugs that shouldn't be caught. We should be able to avoid this by adding an if(isCompile) throw new IllegalStateException().

Having said that, we are currently using IllegalStateException as a sentinal exception, and that is not Daffodil specific. Those may all be used specifically as a sentinal when they are thrown, in which case it should be straightforward to replace them as well. Of course, it is much less likely for a random bug to throw IllegalStateException, so this is less imporant.

Reply:
Yeah, I'd vote for changing it to an IllegalStateException like you mention, and maybe create a new bug to find those usages and change to somethign Daffodil specific so we don't hide actual bugs.

Reply:
NoSuchElementException is a Daffodil exception isn't it? (Maybe we should rename it if that's ambiguous.) It is heavily used by Unparsing as the indicator that an expression which is forward referencing, is not ready to be evaluated yet and should be retried in the future.

At parse time the same exception is used to indicate an expression with a path step for which the corresponding element isn't in the infoset (yet at least - it has to be possible or the expression wouldn't compile, but for whatever reason the parse path didn't put it there when the expression was evaluated at runtime.) 

When doing constant folding, if you touch the infoset at all, you get NoSuchElementException, or so I thought, and that indicates, well, that you were touching the infoset so the expression is non-constant. (Or close enough anyway.)

Reply:
Ah. I see NoSuchElementException is a scala wrapper on a java exception. We should have our own for sure.

Reply:
Nevermind. I am mis-remembering again. Deleting prior comments.

Reply:
Bug filed: https://issues.apache.org/jira/projects/DAFFODIL/issues/DAFFODIL-2176